### PR TITLE
Incremental loading and manual refresh features

### PR DESCRIPTION
**Pull request by aalves08**
Fixes: https://github.com/rancher/dashboard/issues/6540 https://github.com/rancher/dashboard/issues/6498

- Adds the `incremental loading` feature to non-custom list views + workload custom list (and related resources)
- Adds the `manual refresh` feature to non-custom list views + workload custom list (and related resources)
- Adds a new 'Performance' section to Global Settings that allows both features to be enabled and configured

**Works on:**   `DEPLOYMENT`, `CRON_JOB`, `DAEMON_SET`, `JOB`, `STATEFUL_SET`, `REPLICA_SET`, `REPLICATION_CONTROLLER`, `WORKLOAD`, `PODS`, `SECRETS`

Screenshot 2022-07-28 at 19 08 54
Screenshot 2022-07-28 at 19 09 07

### Areas or cases that should be tested
- List views of the following resources: `DEPLOYMENT`, `CRON_JOB`, `DAEMON_SET`, `JOB`, `STATEFUL_SET`, `REPLICA_SET`, `REPLICATION_CONTROLLER`, `WORKLOAD`, `PODS`, `SECRETS`, where these features are applied to.
- Test all the possible combinations: **no** manual refresh **nor** incremental loading + **only** incremental loading + **only** manual refresh + **both** manual refresh and incremental loading
- Be sure to enable the settings on the Performance tab under Global Settings and set an appropriate threshold so that the feature works in the list view you are testing

### Areas which could experience regressions
- Make sure data is loaded from the server and correctly updated on the given list views
- Quickly test other resources to make sure changes introduced did not break the data loading of the other list views
- Check if users with a more restricted access (permissions) to view (or not) these resource are affected

### CODE REVIEW

**Review by richard-cox — CHANGES_REQUESTED** [review]
- From a page where we've sub'd to `PODS` (edit service page) to `PODS` list, the type isn't unwatched
  - I don't think this is important, as the one who's initially subscribed may want to continue watching. Can probably leave this one
- From PODS list where we've not sub'd to `PODS` to Edit Service page, type is not now watched
  - This is more important, as the one that wants to be subscribed never is 
- We should limit this feature to the `EXPLORER` product until we're happier with it (and can ensure it's only enabled for those that use the kube api)
- If the initial partial resource load is executing the manual refresh button should be disabled

File: shell/list/workload.vue [hunk 1]
```diff
@@ -89,6 +95,16 @@ export default {
     },
   },
 
+  // All of the resources that we will load that we need for the loading indicator
+  $loadingResources(route) {
+    const allTypes = route.params.resource === schema.id;
+
+    return {
+      loadResources:     allTypes ? Object.values(WORKLOAD_TYPES) : [route.params.resource],
```
[the remark this answers is not in the document]
Comment (aalves08) [reply shell/list/workload.vue:103]: True. Good one. 👊

File: shell/mixins/resource-fetch.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      // manual refresh vars
+      watch:                      true,
+      isTooManyItemsToAutoUpdate: false,
+      force:                      false,
+      // incremental loading vars
+      incremental:                0,
+      fetchedResourceType:        [],
+    };
+  },
+  beforeDestroy() {
```
[the remark this answers is not in the document]
Comment (aalves08) [reply shell/mixins/resource-fetch.js:36]: That's easy: The mixin is included on both `Workloads` and `ResourceList`, so when you switch list view, both get destroyed (therefore running that hook). I've added a condition to only run on the mixin instance that has been initialized.
On the case of `Workloads` + `ResourceList`, despite the mixin is included on both, it's only initialized on `Workloads` 🙏

File: shell/plugins/dashboard-store/actions.js [hunk 3]
```diff
@@ -75,6 +75,49 @@ export default {
...
+          opt: {
+            ...opt,
+            url: res.pagination?.next
+          }
+        });
+      } else {
+        // We have everything!
+        commit('setHaveAll', { type });
+      }
+    } catch (e) {
```
[the remark this answers is not in the document]
Comment (aalves08) [reply shell/plugins/dashboard-store/actions.js:116]: Removed the whole `try - catch` block

[the remark this answers is not in the document]
Comment (aalves08) [reply shell/plugins/dashboard-store/actions.js:116]: going to resolve this conversation @richard-cox , let me know if you don't agree 🙏

**Review by richard-cox — CHANGES_REQUESTED** [review]
I still see 
- On manual refresh the live columns fail to load (workload restarts & health)

Additionally
- PR description needs updating
  - list of affected resource
  - notes for QA, areas of possible regressions (would be good for them to test permissions)
- On the honker of a cluster, i don't see all resources loading (deployments - pagination shows 2893, side bar shows 3725. pods 2727 vs 2147). This might be the NS filter bug, need to confirm that shows when running the code from `master`

**Review by richard-cox — APPROVED** [review]
If the last changes pass CI then this is good to squash merge.

I've caught the failure to catch failure issue in https://github.com/rancher/dashboard/issues/6574

**Pull request merged**